File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* ptr = head;
        int count = 0;

        // Check if k nodes exists
        while (count < k) {
            if (ptr == NULL) {
                return head;
            }

            ptr = ptr->next;
            count++;
        }

        // Reverse remaining list
        ListNode* prevNode = reverseKGroup(ptr, k);

        // reverse current group
        ptr = head;
        count = 0;

        while (count < k) {
            ListNode* next = ptr->next;
            ptr->next = prevNode;
            prevNode = ptr;
            ptr = next;
            count++;
        }

        return prevNode;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:        
        def merge(left, right):
            head = ListNode(0)
            curr = head
            while left and right:
                if left.val <= right.val:
                    curr.next = left
                    left = left.next
                else:
                    curr.next = right
                    right = right.next
                curr = curr.next
            
            if not left:
                curr.next = right
            
            if not right:
                curr.next = left
            
            return head.next

        def mergeSort(lists, l, r):
            if l > r:
                return lists
            
            if l == r:
                return lists[l]
            
            mid = l + (r - l)//2
            left = mergeSort(lists, l, mid)
            right = mergeSort(lists, mid + 1, r)
            return merge(left, right)
        
        if not lists:
            return None
        
        return mergeSort(lists, 0, len(lists) - 1)

        